packages/@aws-cdk/cdk/lib/cloudformation/resource.ts

```diff
@@ -213,28 +219,36 @@ export class Resource extends Referenceable {
     return properties;
   }
 
-  private renderDependsOn() {
-    const logicalIDs = new Set<string>();
-    for (const d of this.dependsOn) {
-      addDependency(d);
+  /**
+   * Final preparation before rendering.
+   *
+   * Take all dependencies, find the CloudFormation Resources in them,
+   * and take a dependency on their logical IDs. If the discovered
+   * resources turn out to be in different Stacks, take a Stack dependency
+   * instead.
+   */
+  protected prepare() {
+    super.prepare();
+
+    // As an optimization, do the stack dependencies first on the parents
```

Comment:
Would it make more sense perhaps to do this globally at the stack instead of distribute to each resource?